### PR TITLE
Rename typing._collect_parameters

### PR DESCRIPTION
**Pull request by JelleZijlstra**
function. Unfortunately, released versions of typing_extensions
monkeypatch this function without the extra parameter, which makes
it so things break badly if current main is used with typing_extensions.

```
% ~/py/cpython/python.exe test_typing_extensions.py
Traceback (most recent call last):
  File "/Users/jelle/py/typing_extensions/src/test_typing_extensions.py", line 42, in <module>
    from _typed_dict_test_helper import Foo, FooGeneric, VeryAnnotated
  File "/Users/jelle/py/typing_extensions/src/_typed_dict_test_helper.py", line 17, in <module>
    class FooGeneric(TypedDict, Generic[T]):
                                ~~~~~~~^^^
  File "/Users/jelle/py/cpython/Lib/typing.py", line 431, in inner
    return func(*args, **kwds)
  File "/Users/jelle/py/cpython/Lib/typing.py", line 1243, in _generic_class_getitem
    return _GenericAlias(cls, args)
  File "/Users/jelle/py/cpython/Lib/typing.py", line 1420, in __init__
    self.__parameters__ = _collect_parameters(
                          ~~~~~~~~~~~~~~~~~~~^
        args,
        ^^^^^
        enforce_default_ordering=enforce_default_ordering,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
TypeError: _collect_parameters() got an unexpected keyword argument 'enforce_default_ordering'
```

Fortunately, the monkeypatching is not needed on Python 3.13, because CPython
now implements PEP 696. By renaming the function, we prevent the monkeypatch
from breaking typing.py internals.

### CODE REVIEW

**Comment by AlexWaygood** [comment]
There are some uses in the wild... https://github.com/wyfo/apischema/blob/d48cc010c417e9c49db25b7a8683621b3c305b44/apischema/typing.py#L59-L62

And we could also fix it over at `typing_extensions with this diff:

```diff
--- a/src/typing_extensions.py
+++ b/src/typing_extensions.py
@@ -2836,7 +2836,8 @@ else:
 
         return tuple(parameters)
 
-    typing._collect_parameters = _collect_parameters
+    if not _PEP_696_IMPLEMENTED:
+        typing._collect_parameters = _collect_parameters
```

It's true that this change has the advantage that already-released versions of `typing_extensions` are fixed as well to-be-released versions of `typing_extensions`. But I think already-released versions of `typing_extensions` are badly broken anyway on Python 3.13, due to the error pointed out in https://github.com/python/typing_extensions/issues/377#issue-2289419762

**Comment by JelleZijlstra** [comment]
> But I think already-released versions of `typing_extensions` are badly broken anyway on Python 3.13, due to the error pointed out in [python/typing_extensions#377 (comment)](https://github.com/python/typing_extensions/issues/377#issue-2289419762)

I think that error shows up only if you use `typing_extensions.TypeVar` though, while the error this PR fixes breaks any use of Generic after `typing_extensions` is imported, which feels much worse.



> There are some uses of this function in the wild, which this change would break... https://github.com/wyfo/apischema/blob/d48cc010c417e9c49db25b7a8683621b3c305b44/apischema/typing.py#L59-L62

I think we could address this by keeping an alias `_collect_parameters = _collect_type_parameters` (or a wrapper function that raises DeprecationWarning).

**Comment by AlexWaygood** [comment]
Okay, sounds reasonable.

> I think we could address this by keeping an alias `_collect_parameters = _collect_type_parameters` (or a wrapper function that raises DeprecationWarning).

Or we could do this in the `__getattr__` method down at the bottom? (With `do_the_deprecation_warning()` replaced by the appropriate logic to do a deprecation warning)

```diff
--- a/Lib/typing.py
+++ b/Lib/typing.py
@@ -3770,6 +3770,9 @@ def __getattr__(attr):
     elif attr in {"ContextManager", "AsyncContextManager"}:
         import contextlib
         obj = _alias(getattr(contextlib, f"Abstract{attr}"), 2, name=attr, defaults=(bool | None,))
+    elif attr == "_collect_parameters":
+        do_the_deprecation_warning()
+        return _collect_type_parameters
     else:
         raise AttributeError(f"module {__name__!r} has no attribute {attr!r}")
     globals()[attr] = obj
```

File: Lib/typing.py [hunk 1]
```diff
@@ -3770,6 +3770,16 @@ def __getattr__(attr):
...
         obj = _alias(getattr(contextlib, f"Abstract{attr}"), 2, name=attr, defaults=(bool | None,))
+    elif attr == "_collect_parameters":
+        import warnings
+
+        depr_message = (
+            "The private _collect_parameters function is deprecated and will be"
+            " removed in a future version of Python. Any use of private functions"
+            " is discouraged and may break in the future."
+        )
+        warnings.warn(depr_message, category=DeprecationWarning, stacklevel=1)
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
        warnings.warn(depr_message, category=DeprecationWarning, stacklevel=2)
```

Or it literally displays line 3781 of `typing.py` in the DeprecationWarning message as the line that's doing the problematic thing when you trigger the DeprecationWarning:

```pycon
Python 3.14.0a0 (heads/main:a895756aec, May 10 2024, 14:25:23) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import _collect_parameters
/Users/alexw/dev/cpython/Lib/typing.py:3781: DeprecationWarning: The private _collect_parameters function is deprecated and will be removed in a future version of Python. Any use of private functions is discouraged and may break in the future.
  warnings.warn(depr_message, category=DeprecationWarning, stacklevel=1)
>>> 
```

You can add a test for this by doing

```py
with self.assertWarns(DeprecationWarning) as cm:
    typing._collect_parameters

self.assertEqual(cm.filename, __file__)

Comment (JelleZijlstra) [reply Lib/typing.py]: Thanks, should have looked at the warning more.

File: Lib/test/test_typing.py [hunk 2]
```diff
@@ -6350,6 +6352,12 @@ def test_deprecation_for_no_type_params_passed_to__evaluate(self):
 
         self.assertEqual(cm.filename, __file__)
 
+    def test_collect_parameters(self):
+        typing = import_helper.import_fresh_module("typing")
+        with self.assertWarns(DeprecationWarning) as cm:
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]: I would probably do this just to make sure that it is actually _this_ function being deprecated, not some other function:

```suggestion
        with self.assertWarnsRegex(
            DeprecationWarning,
            "The private _collect_parameters function is deprecated"
        ) as cm:
```

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Pull request merged**